### PR TITLE
sinkv2(ticdc): improve sinkv2 progress tracker

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Issue Number: ref #5928 

### What is changed and how it works?

Improve performance about sinkv2 progress tracker.

#### Tests 

 - Unit test
 - Manual test
    图片

    The manual test shows progress tracker won't be the bottleneck anymore.

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/sinkv2/tablesink/progress_tracker.go [hunk 1]
```diff
@@ -15,10 +15,11 @@ package tablesink
 
 import (
 	"context"
+	"math"
 	"sync"
+	stdAtomic "sync/atomic"
```
Comment (hi-rustin) [inline cdc/sinkv2/tablesink/progress_tracker.go:20]:
```suggestion
	stdatomic "sync/atomic" or satomic "sync/atomic" 
```
We seem to use very little camel naming in package names.

File: cdc/sinkv2/tablesink/progress_tracker.go [hunk 2]
```diff
@@ -44,109 +54,170 @@ const (
...
-	defer r.lock.Unlock()
-	r.pendingEventAndResolvedTs.Put(key, nil)
+// addEvent is used to add the pending event key. `postEventFlush` should be called
+// when the event has been flushed.
+func (r *progressTracker) addEvent() (postEventFlush func()) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.nextEventID += 1
+	eventID := r.nextEventID - 1
```
Comment (hi-rustin) [inline cdc/sinkv2/tablesink/progress_tracker.go:109]: This code looks weird. Why not get eventID first?

File: cdc/sinkv2/tablesink/progress_tracker.go [hunk 3]
```diff
@@ -44,109 +54,170 @@ const (
...
+		}
 	}
+
+	return r.lastMinResolvedTs
 }
 
 // minTs returns the last min resolved ts.
 // This means that all data prior to this point has already been processed.
 // It can be considered as CheckpointTs of the table.
 // Notice: must hold the lock.
```
Comment (hi-rustin) [inline cdc/sinkv2/tablesink/progress_tracker.go:210]: Maybe we can remove this.

File: cdc/sinkv2/tablesink/progress_tracker.go [hunk 4]
```diff
@@ -44,109 +54,170 @@ const (
...
+	resolvedTsCache []pendingResolvedTs
+
+	lastMinResolvedTs model.ResolvedTs
 }
 
 // newProgressTracker is used to create a new progress tracker.
 // The last min resolved ts is set to 0.
 // It means that the table sink has not started yet.
-func newProgressTracker(tableID model.TableID) *progressTracker {
+func newProgressTracker(tableID model.TableID, bufferSize uint64) *progressTracker {
```
Comment (hi-rustin) [inline cdc/sinkv2/tablesink/progress_tracker.go:87]: When to change the buffer size?

**Review by hi-rustin — COMMENTED** [review]
Do we have any docs about this algorithm? It's difficult to understand for me.

**Comment by hi-rustin** [comment]
The lint failed. https://ci.pingcap.net/blue/organizations/jenkins/lint-check/detail/lint-check/15220/pipeline/

File: cdc/sinkv2/tablesink/progress_tracker.go (hunk 4 above)
Comment (hicqu) [reply cdc/sinkv2/tablesink/progress_tracker.go:87]: Only tests will change it.

**Comment by hicqu** [comment]
> Do we have any docs about this algorithm? It's difficult to understand for me.

I have commentted it on `progressTracker`. Just use `pendingEvents` as a lock-free bit-map.

**Review by hi-rustin — APPROVED** [review]
Looks great!

**Comment by hi-rustin** [comment]
I added some comments when I read the code. You can add it if you wish, it may be helpful to read the code.

<details><summary>Click me to see diff</summary>

```diff
diff --git a/cdc/sinkv2/tablesink/progress_tracker.go b/cdc/sinkv2/tablesink/progress_tracker.go
index cde6b71c5..a20e5c31e 100644
--- a/cdc/sinkv2/tablesink/progress_tracker.go
+++ b/cdc/sinkv2/tablesink/progress_tracker.go
@@ -33,7 +33,6 @@ const (
 	waitingInterval = 100 * time.Millisecond
 	// warnDuration is the duration to warn the progress tracker is not closed.
 	warnDuration = 3 * time.Minute
-
 	// A progressTracker contains several internal fixed-length buffers.
 	defaultBufferSize uint64 = 1024 * 1024
 )
@@ -69,7 +68,7 @@ type progressTracker struct {
 	// Following fields are protected by `mu`.
 	mu sync.Mutex
 
-	// Used to generated the next eventID.
+	// Used to generate the next eventID.
 	nextEventID uint64
 
 	// Every received event is a bit in `pendingEvents`.
@@ -115,17 +114,24 @@ func (r *progressTracker) addEvent() (postEventFlush func()) {
 	r.nextEventID += 1
 
 	bufferCount := len(r.pendingEvents)
+	// The buffer is full or empty, we need to allocate a new buffer.
 	if bufferCount == 0 || (uint64(len(r.pendingEvents[bufferCount-1])) == r.bufferSize && bit == 0) {
 		buffer := make([]uint64, 0, r.bufferSize)
 		r.pendingEvents = append(r.pendingEvents, buffer)
 		bufferCount += 1
 	}
 
+	// Set the bit to 0.
 	if bit == 0 {
 		r.pendingEvents[bufferCount-1] = append(r.pendingEvents[bufferCount-1], 0)
 	}
 	lastBuffer := r.pendingEvents[bufferCount-1]
 
+	// Set the corresponding bit to 1.
+	// For example, if the eventID is 3, the bit is 3 % 64 = 3.
+	// 0000000000000000000000000000000000000000000000000000000000000000 ->
+	// 0000000000000000000000000000000000000000000000000000000000001000
+	// When we advance the progress, we can try to find the first 0 bit to indicate the progress.
 	postEventFlush = func() { stdatomic.AddUint64(&lastBuffer[len(lastBuffer)-1], 1<<bit) }
 	return
 }
@@ -153,29 +159,40 @@ func (r *progressTracker) advance() model.ResolvedTs {
 
 	// `pendingEvents` is like a 3-dimo bit array. To access a given bit in the array,
 	// use `pendingEvents[idx1][idx2][idx3]`.
+	// The first index used to access the buffer.
+	// The second index used to access the uint64 in the buffer.
+	// The third index used to access the bit in the uint64.
 	offset := r.nextToResolvePos - r.nextToReleasePos
 	idx1 := offset / (r.bufferSize * 64)
 	idx2 := offset % (r.bufferSize * 64) / 64
 	idx3 := offset % (r.bufferSize * 64) % 64
 
 	for {
+		// All events are resolved.
 		if r.nextToResolvePos >= r.nextEventID {
-			// All events are resolved.
 			break
 		}
 
 		currBitMap := stdatomic.LoadUint64(&r.pendingEvents[idx1][idx2])
+		// All bits in the currBitMap are 1.
 		if currBitMap == math.MaxUint64 {
+			// Move to the next bit map.(aka next uint64)
 			idx2 += 1
+			// Move to the next buffer.
 			if idx2 >= r.bufferSize {
 				idx2 = 0
 				idx1 += 1
 			}
+			// Move forward the nextToResolvePos with 64.
+			// Because all bits in the currBitMap are 1.
 			r.nextToResolvePos += 64 - idx3
+			// Move to the first bit.
 			idx3 = 0
 		} else {
+			// Try to find the first 0 bit.
 			for i := idx3; i < 64; i++ {
 				if currBitMap&uint64(1<<i) == 0 {
+					// Move forward the nextToResolvePos with `i`.
 					r.nextToResolvePos += i - idx3
 					break
 				}
@@ -184,9 +201,12 @@ func (r *progressTracker) advance() model.ResolvedTs {
 		}
 	}
 
+	// We got some resolved ts to update.
 	if r.nextToResolvePos > 0 {
 		for len(r.resolvedTsCache) > 0 {
 			cached := r.resolvedTsCache[0]
+			// Make sure the resolved ts is not greater than the nextToResolvePos.
+			// And the last min resolved ts.
 			if cached.offset <= r.nextToResolvePos-1 {
 				r.lastMinResolvedTs = cached.resolvedTs
 				r.resolvedTsCache = r.resolvedTsCache[1:]
@@ -199,6 +219,7 @@ func (r *progressTracker) advance() model.ResolvedTs {
 		}
 	}
 
+	// If we finished one buffer, we can release it.
 	for r.nextToResolvePos-r.nextToReleasePos >= r.bufferSize*64 {
 		r.nextToReleasePos += r.bufferSize * 64
 		r.pendingEvents = r.pendingEvents[1:]

```

</details>

**Comment by hi-rustin** [comment]
Thanks for working on this!  💚 💙 💜 💛 ❤️

**Review by nongfushanquan — APPROVED** [review]

**Pull request merged**